Refactoring: Bookmarks as Fragment

stefano-cliqz opened the pull request:
1. Incognito mode in another process
2. Bookmarks as a Fragment using Otto

I need also to refactor tabs management in a fragment so expect another pull request soon.

anthonycr commented:
Looks awesome, going to review.

anthonycr commented on external/netcipher:
```diff
@@ -1 +1 @@
-Subproject commit be233148c1c644b77b4f90089c82d85a50940c57
+Subproject commit 442d282b07940a9992f54fcf61c44e12472cca83
```
Could you run an update on the netcipher library? The commit that's being reverted to creates some problems with proxying. You can update by running:
```
git submodule foreach git reset --hard
git submodule update --remote
```

anthonycr commented on app/src/main/java/acr/browser/lightning/utils/Utils.java:
```diff
@@ -309,4 +312,20 @@ public static void drawTrapezoid(Canvas canvas, int color, boolean withShader) {
+     * @param str   the string to hash
+     * @return      the long representation of the hash
+     */
+    public static long hash(final String str) {
```
+1 much better than relying on string.hashCode()

kuc replied:
...but practically swallowing an exception - not so much. This probably should "never happens", but I would rethrow RuntimeException instead of nearly silently generating many hash collisions.

stefano-cliqz replied:
@kuc you are right, I'll change this

anthonycr commented on app/src/main/java/acr/browser/lightning/activity/BrowserActivity.java:
```diff
@@ -2494,12 +2101,13 @@ public void longClickPage(final String url) {
                     longPressHistoryLink(newUrl);
                 }
             } else if (currentUrl.endsWith(BookmarkPage.FILENAME)) {
+                /* TODO Wtf is this?
```
If you go into settings and change your homepage to bookmarks, a bookmark page will be displayed in place of the home screen. The bookmark logic needs to handle all bookmark long presses, edits, and deletions for the HTML page as well as the bookmark drawer.

stefano-cliqz replied:
Working on it: I have to extract the dialogs from the BookmarksFragment to support this.

anthonycr replied:
@stefano-cliqz cool, I'll hold off on making other changes so that I don't make your time harder with more merge conflicts, as I see a recent change I made inadvertently created a conflict.

anthonycr commented on app/src/main/java/acr/browser/lightning/database/BookmarkManager.java:
```diff
@@ -251,24 +257,36 @@ public synchronized void exportBookmarks(Activity activity) {
     public synchronized List<HistoryItem> getAllBookmarks(boolean sort) {
-        List<HistoryItem> bookmarks = new ArrayList<>();
-        File bookmarksFile = new File(mContext.getFilesDir(), FILE_BOOKMARKS);
+		final List<HistoryItem> bookmarks = new ArrayList<>();
```
Please use 4 spaces instead of a tab

DF1E commented:
Maybe you can a this feature: #287

anthonycr commented on app/src/main/java/acr/browser/lightning/activity/BrowserActivity.java:
```diff
@@ -358,38 +342,11 @@ private synchronized void initialize() {
 
             @Override
             public void run() {
-                mBookmarkManager = BookmarkManager.getInstance(mActivity.getApplicationContext());
```
Removing initialization of the BookmarkManager causes a NPE when opening the bookmarks page (settings > general settings > homepage > bookmarks)

anthonycr commented on app/src/main/res/layout/activity_main.xml:
```diff
@@ -27,7 +27,19 @@
+        <FrameLayout
+            android:weightSum="1"
+            android:layout_gravity="end"
+            android:fitsSystemWindows="true"
```
For some reason, the drawer is not visible behind the status bar - I think this can be fixed by adding the fitsSystemWindows flag to the fragment tag as well. Also, the drawer shadow is no longer visible.

stefano-cliqz replied:
The problem was the FrameLayout background (transparent)

stefano-cliqz commented:
I fixed the BookmarkPage.
In the meanwhile, I added also Dagger 2 for dependencies injection. This permits to have more modularized code. Hope this is fine with you.
Waiting for your comments.

anthonycr commented:
Reviewing the updates now. I'm not familiar with Dagger, so right now I think it's fine to add, but I'll let you know if I decide otherwise.

stefano-cliqz commented:
The difference in terms of lines it's not so huge (something around 300 lines): if you remove comments, imports (duplicated due to the refactoring) and ceremony only few lines have been added. Many of them are very simple (like the classes used by the Bus).

anthonycr commented:
@stefano-cliqz good point :)

anthonycr commented on app/src/main/java/acr/browser/lightning/fragment/BookmarksFragment.java:
```diff
@@ -0,0 +1,361 @@
+    public void bookmarkDeleted(final BookmarkEvents.Deleted event) {
+        final HistoryItem item = event.item;
+        final int size = mBookmarks.size();
+        mBookmarks.remove(event);
```
````item```` should be removed from mBookmarks rather than ````event````

anthonycr commented:
Looks good, merging.
